user_tools/src/spark_rapids_tools/tools/qualification_stats_report.py

```diff
@@ -0,0 +1,162 @@
+        # Convert durations from milliseconds to seconds
+        self.unsupported_operators_df[['Stage Duration', 'App Duration']] /= 1000
+        self.stages_df[['Stage Task Duration', 'Unsupported Task Duration']] /= 1000
+
+    def _preprocess_dataframes(self) -> None:
+        self.logger.info('Preprocessing dataframes...')
+
+        # Filter out 'WholeStageCodegen' operators as the child operators are already included
+        # in the other rows
+        self.execs_df = self.execs_df[
+            ~self.execs_df['Exec Name'].str.startswith('WholeStageCodegen')]
+
+        # Split 'Exec Stages' and explode the list into separate rows so that the stageID
+        # from this dataframe can be matched with the stageID of stages dataframe
+        self.execs_df['Exec Stages'] = self.execs_df['Exec Stages'].str.split(':')
+        self.execs_df = self.execs_df.explode('Exec Stages').dropna(subset=['Exec Stages'])
+        self.execs_df['Exec Stages'] = self.execs_df['Exec Stages'].astype(int)
+
+        # Remove duplicate 'Stage ID' rows and rename some columns so that join on dataframes
+        # can be done easily
+        self.stages_df = self.stages_df.drop_duplicates(subset=['App ID', 'Stage ID'])
+        self.stages_df.rename(columns={'Stage Task Duration': 'StageTaskDuration'}, inplace=True)
+        self.execs_df.rename(columns={'Exec Name': 'Operator'}, inplace=True)
+        self.unsupported_operators_df.rename(columns={'Unsupported Operator': 'Operator'},
+                                             inplace=True)
+        self.logger.info('Preprocessing dataframes completed.')
+
+    def _merge_dataframes(self) -> None:
+        self.logger.info('Merging dataframes to get stats...')
+        self._preprocess_dataframes()
+
+        # Merge execs_df with stages_df
+        merged_df = self.execs_df.merge(self.stages_df, left_on=['App ID', 'Exec Stages'],
+                                        right_on=['App ID', 'Stage ID'], how='left')
+
+        # Merge with unsupported_operators_df to find unsupported operations
+        merged_df = merged_df.merge(self.unsupported_operators_df,
+                                    on=['App ID', 'SQL ID', 'Stage ID', 'Operator'],
+                                    how='left', indicator=True)
+        merged_df['Supported'] = merged_df['_merge'] == 'left_only'
+        merged_df.drop(columns=['_merge', 'Exec Stages'], inplace=True)
+
+        # Calculate total duration by summing unique stages per SQLID
+        total_duration_df = merged_df.drop_duplicates(subset=['App ID', 'SQL ID', 'Stage ID']) \
+            .groupby(['App ID', 'SQL ID'])['StageTaskDuration'] \
+            .sum().reset_index().rename(columns={'StageTaskDuration': 'TotalSQLDuration'})
+
+        merged_df = merged_df.merge(total_duration_df, on=['App ID', 'SQL ID'], how='left')
+
+        # Mark unique stage task durations
+        merged_df['Unique StageTaskDuration'] = ~merged_df.duplicated(
```

Comment:
so for each operator we will get all stages and then possibly separate rows if have both supported and unsupported and we just want to make sure we have those unique so we aren't double counting on time, correct?

So does this mean if you have an two operators in a stage, one that is supported and one that is not supported that we the time will be in there twice, correct?  which is fine I just want to make sure I'm understanding properly.

Reply:
Thanks! I updated it to handle this case. Until now, it was taking into account same operators within a SQL ID and the assumption was that the operator will be either supported or unsupported within a stage. But there could be a scenario where the same operator can be supported and unsupported within a stage(due to underlying Expression not supported).

Sample output: There are 4 entries of Project for SQLID=1, StageID(5,6) in execs.csv and 3 are unsupported as below:
unsupported_operators_df

```
    App ID  SQL ID  Stage ID       Operator
0       1       3         7  HashAggregate
1       1       1         6        Project
2       2       1        -1        Project
3       1       1        -1         Filter
4       1       1         5        Project
5       1       1         5        Project

```
stages_df:
```
 stages_df
   App ID  Stage ID  StageTaskDuration
0       1         5                 50
2       1         4                 20
3       1         3                100
4       1         6                 60
5       1         7                100

```

final dataframe output(before renaming columns):
```
      App ID  SQL ID     Operator      Count  StageTaskDuration  TotalSQLDuration  % of Total SQL Duration  Supported
0       1       1         Filter         2                70                230                30.434783       True
1       1       1        Project         3                110               230                47.826087      False
2       1       1        Project         1                 50               230                21.739130       True
3       1       1           Sort         3                170               230                73.913043       True
4       1       3  HashAggregate         1                100               100               100.000000      False

```

Reply:
so I assume the sql id 1 above example has some execs - likely the Project that are unsupposed in the same stages as another exec that is supported, correct?  Because you hadd up the 4 of those 70 + 110 + 50 + 170 and its more then the 230 for total.